### ISSUE TITLE
Downstream skip SSL/certificate validation per configuration

### ISSUE THREAD

**Issue by Burnaster** [ThreeMammals/Ocelot#309]
## Expected Behavior / New Feature
Using ocelot in a development/test environment without valid certificates / self signed certificates failes with:
System.Net.Http.CurlException: Peer certificate cannot be authenticated with given CA certificates

## Actual Behavior / Motivation for New Feautre
It would be nice to turn off SSL/certificate validation eather per route or globally.

```{
  "ReRoutes": [
    {
      "DownstreamPathTemplate": "/api/people",
      "DownstreamScheme": "http",
      "DownstreamSkipSslValidation": true,
      "DownstreamHostAndPorts": [
        {
          "Host": "localhost",
          "Port": 8238
        }
      ],
      "UpstreamPathTemplate": "/api/people",
      "UpstreamHttpMethod": [ "Post" ]
    }
  ],
  "GlobalConfiguration": {
    "BaseUrl": "http://localhost:18970",
    "SkipSslValidation": true
  }
}
```
Which would lead to the setup of the `HttpClientHandler` in `Ocelot.Requester.HttpClientBuilder`.

```
if (request.DownstreamReRoute.HttpHandlerOptions.SkipSslValidation)
{
    _httpclientHandler.ServerCertificateCustomValidationCallback = (message, cert, chain, errors) => true;
}
```

## Steps to Reproduce the Problem

  1. use a downstream https route that uses a invalid certificate.

## Specifications

  - Version: latest

**Comment by TomPallister** [comment]
@Burnaster Thanks for your interest in the project. I think you can just do this in your startup.cs / program.cs

```csharp
ServicePointManager.ServerCertificateValidationCallback += (sender, certificate, chain, sslPolicyErrors) => true;
```

let me know if that doesn't help!

**Comment by Burnaster** [comment]
@TomPallister Could it be that the behaviour on linux is different?

Error making http request, exception: System.Net.Http.HttpRequestException: An error occurred while sending the request. ---> System.Net.Http.CurlException: SSL connect error
at System.Net.Http.CurlHandler.ThrowIfCURLEError(CURLcode error)
...

**Comment by TomPallister** [comment]
@Burnaster it looks like it is supported [here](https://docs.microsoft.com/en-us/dotnet/api/system.net.servicepointmanager.servercertificatevalidationcallback?view=netcore-2.0). Does it not work for you?

**Comment by Burnaster** [comment]
@TomPallister unfortunately it did not. But also the validation callback on the HttpClientHandler did not work. Im still a bit confused with the use of dotnet core on linux and cURL to handle HttpClient requests. 

To not be blocked, i use http on the ocelot api gateway. But i will update this issue if i find out something worth shareing.

**Comment by TomPallister** [comment]
@Burnaster try this article https://www.humankode.com/asp-net-core/develop-locally-with-https-self-signed-certificates-and-asp-net-core

Self signed should work, I guess you need to get your computer to trust the certificate still.

**Comment by handashuai** [comment]
我也想问，怎么加载 https

**Comment by TomPallister** [comment]
Maybe the new 2.1 handler will fix this issue.

**Comment by TomPallister** [comment]
@Burnaster btw did you ever test to see if this works when calling your service without Ocelot? E.g. just using a console app using HttpClient for example?

**Comment by Burnaster** [comment]
@TomPallister no i did not test that. as mentioned earlier i plan to do additional tests on multiple platforms/frameworks.

**Comment by TomPallister** [comment]
@Burnaster cool, might be worth test to see if this is an Ocelot problem, I will have a try myself when I get a chance!

**Comment by Burnaster** [comment]
@TomPallister I did some early tests. I executed my test program (see below) on a windows 10 64bit and on a cloud foundry app (cflinuxfs2: see https://github.com/cloudfoundry/cflinuxfs2) to call a untrusted HTTPS with a GET request. One time using the ServerCertificateCustomValidationCallback on the  ServicePointManager callback and one time using it on a HttpClientHandler.

System | Method | Outcome
------------ | ------------- | -------------
Win64 .net core 2.0.6 | Handler | 200=OK
Win64 .net core 2.0.6 | SPM | Error: System.Net.Http.HttpRequestException: An error occurred while sending the request. ---> System.Net.Http.WinHttpException: Es ist ein Sicherheitsfehler aufgetreten
cflinuxfs2 .net core 2.0.6 | Handler | 200 = OK
cflinuxfs2 .net core 2.0.6 | SPM | System.Net.Http.HttpRequestException: An error occurred while sending the request. ---> System.Net.Http.CurlException: Peer certificate cannot be authenticated with given CA certificates

Next would to test the same on dotnet core 2.1, the release should be close so i think i wait and do the tests with the release instead of a preview. What would also be interessting is to repeat on a different linux. As far as i understood the behavior is based on the cURL build (with or without openssl, etc.) used in the particular distro. 

Can you reproduce my results?

```csharp
using System;
using System.Net;
using System.Net.Http;
using System.Threading.Tasks;

namespace HttpClientSkipSsl
{
    class Program
    {
        private const string RequestUri = "https://untrustedurl";

        static void Main()
        {
            RunAsync().GetAwaiter().GetResult();
        }

        private static async Task<HttpStatusCode> GetWithHttpHandler()
        {
            using (var httpHandler = new HttpClientHandler())
            {
                httpHandler.ServerCertificateCustomValidationCallback = (mst, crt, chn, plc) => true;
                var http = new HttpClient(httpHandler);
                var response = await http.GetAsync(RequestUri);

                return response.StatusCode;
            }
        }

        private static async Task<HttpStatusCode> GetWithSpm()
        {
            ServicePointManager.ServerCertificateValidationCallback += (sender, certificate, chain, sslPolicyErrors) => true;

            var http = new HttpClient();
            var response = await http.GetAsync(RequestUri);

            return response.StatusCode;
        }

        private static async Task RunAsync()
        {
            try
            {
                var statusWithHandler = await GetWithHttpHandler();

                Console.WriteLine($"Get invalid HTTPS url with HttpHandler status code: {statusWithHandler}");
            }
            catch (Exception exception)
            {
                Console.WriteLine($"Get invalid HTTPS url with HttpHandler error: {exception.Message}");
            }

            try
            {
                var statusWithHandler = await GetWithSpm();

                Console.WriteLine($"Get invalid HTTPS url with SPM status code: {statusWithHandler}");
            }
            catch (Exception exception)
            {
                Console.WriteLine($"Get invalid HTTPS url with SPM error: {exception}");
            }
        }
    }
}
```

**Comment by TomPallister** [comment]
@Burnaster thanks for this testing. This is also what I see. I need to expose the handler in some way from Ocelot for it to work with I trusted SSL. Though I’m not sure if giving that option is a good idea. What do you think? Is it not possible to get your certificate trusted so SSL works normally?

**Comment by TomPallister** [comment]
@Burnaster I have done a quick change for ignoring SSL warnings, you can see it [here](https://github.com/ThreeMammals/Ocelot/pull/325) if you want to make sure it works. I have tested and am confident it does ;) There are even some acceptance tests for it now!

**Comment by TomPallister** [comment]
@Burnaster I have released this change in version 5.5.5 so you should be able to ignore ssl errors now. Let me know if you still have problems.
